### ISSUE TITLE
[Proposal] Attaching metadata to schema

### ISSUE THREAD

**Issue by Florian3k** [vriad/zod#158]
I am proposing to add `.meta()` method that allows attaching metadata to schema

## Proposed way it would work
Basic usage:
```ts
const schema = z.string().meta({ some: "data" })

schema.meta // => { some: "data" }
```
Multiple `.meta` calls would merge (shallowly) with previous metadata and overwrite repeated fields:
```ts
const schema = z.string()
  .meta({ foo: 1, bar: 2, nested: { prop: "data" }})
  .meta({ baz: 3, nested: { otherProp: "data" }})
  .meta({ foo: 4 })

schema.meta // => { foo: 4, bar: 2, baz: 3, nested: { otherProp: "data" }}
````
`.meta` would have following signature:
```ts
meta(metadata: Record<string, any>): this
```
## Use cases:
One of the uses cases i can think of is generating forms from schema. Imagine library that does that, more or less in following way:
```jsx
// from this schema
const schema = z.object({
  username: z.string(),
  age: z.number().min(18),  
})
// this is generated form:
<form>
  <input type="text" name="username" />
  <input type="number" name="age" />
</form>
```
Now, let's say we want to use certain component (assuming React here) for certain field. For example we want to use `FancyPhoneNumberField` component for `phone` field:
```jsx
// from this schema
const schema = z.object({
  username: z.string(),
  age: z.number().min(18),
  phone: z.string().meta({ formgenComponent: FancyPhoneNumberField })
})
// this is generated form:
<form>
  <input type="text" name="username" />
  <input type="number" name="age" />
  <FancyPhoneNumberField name="phone" />
</form>
```
_Of course library like this would probably add much more props but this is minimal version to understand use case_

There are probably many other cases where attaching metadata would be useful.

 I am willing to implement and document this if gets accepted

**Comment by vriad** [comment]
Not a fan of this for reasons explained in #76 and #138.

tldr: if you want to define a data structure containing metadata for your form fields, Zod should be a component within it, not the container for it. You would get stronger typing using an approach like this: https://github.com/vriad/zod/issues/76#issuecomment-673271898

**Comment by Florian3k** [comment]
Haven't seen these two issues before. I think this one can be closed.

**Comment by jedwards1211** [comment]
Here's how I accomplish this: https://github.com/colinhacks/zod/issues/76#issuecomment-1887838005
